js/server/types/User.js

```diff
@@ -0,0 +1,228 @@
+  constructor () {
+    this.tickets = [];
+    this.claimableRewardsOnWithdrawnAssets = 0;
+    this.dispensed = 0;
+    this.forfeited = 0;
+    this.currentTotalClaimableReward = 0;
+    this.reservedReward = 0;
+    this.totalTicketsAmountSum = 0;
+    this.claimableCommissionsOnDelegatorRewards = 0;
+    this.totalRewardAtMaturity = 0;
+    this.ticketAmountAtMaturity = 0;
+    this.yieldAtMaturity = 0;
+    this.nextRewardShare = 0;
+    this.currentYieldOnTickets = 0;
+    // dates
+    this.maturityDate = undefined;
+    this.maturityDateISO = undefined;
+    this.yearsToMaturity = undefined;
+    this.currentAPYOnTickets = undefined;
+    this.maturityDateMs = undefined;
+    // timevalues
+    this.futureReward = undefined;
+    this.nextReward = undefined;
+    this.nextRewardProjectedFutureReward = undefined;
+    this.nextRewardProjectedAPYOnTickets = undefined;
+  }
+
+  updateMaturityTimeProps (userAtFinalTimestamp, currentTimestampInMs) {
+    this.maturityDate = userAtFinalTimestamp.maturityDate;
+    this.maturityDateISO = userAtFinalTimestamp.maturityDateISO;
+    const msToMaturity =
+      userAtFinalTimestamp.maturityDateMs - currentTimestampInMs;
+    this.yearsToMaturity = msToMaturity / 1000 / 60 / 60 / 24 / 365;
+    this.currentAPYOnTickets =
+      this.currentYieldOnTickets / this.yearsToMaturity;
+  }
+
+  cloneWith (props) {
+    let next = new User();
+    Object.assign(next, this);
+    Object.assign(next, props);
+    return next;
+  }
+
+  refreshTotalTicketAmountSum () {
+    this.totalTicketsAmountSum = _.sum(this.tickets.map(t => t.amount));
+  }
+
+  updateRewards (timestampTicketsAmountSum) {
+    this.refreshTotalTicketAmountSum();
+    this.currentTotalClaimableReward =
+      this.claimableRewardsOnWithdrawnAssets +
+      _.sum(this.tickets.map(t => t.reward * t.mul));
+    this.reservedReward =
+      this.claimableRewardsOnWithdrawnAssets +
+      _.sum(this.tickets.map(t => t.reward));
+    this.nextRewardShare =
+      this.totalTicketsAmountSum / timestampTicketsAmountSum;
+  }
+
+  updateUserMaturityDates (
+    userAtPrevTimestamp,
+    isAfterRewardPeriod,
+    currentTimestampInMinutes,
+    nextBucketGlobalReward
+  ) {
+    let maturityDate = userAtPrevTimestamp.maturityDate;
+    let maturityDateISO = userAtPrevTimestamp.maturityDateISO;
+    let maturityDateMs = userAtPrevTimestamp.maturityDateMs;
+    let maturityDateMoment;
+    if (
+      maturityDate === undefined && // maturity date not yet reached
+      isAfterRewardPeriod && // reward period is over
+      this.currentTotalClaimableReward === this.reservedReward // rewards have matured
+    ) {
+      maturityDateMoment = moment
+        .utc(config.START_DATETIME)
+        .add(currentTimestampInMinutes, 'm');
+      maturityDate = maturityDateMoment.format('MMMM Do YYYY, h:mm:ss a');
+      maturityDateMs = maturityDateMoment.valueOf();
+      maturityDateISO = this.maturityDateISO = maturityDateMoment.toISOString();
```

Comment:
Double assign to `this` here

Reply:
On purpose. Also from original code in `process-vs`. Could definitely put that on a new line to improve clarity though

Reply:
It gets re-assigned to this 2 lines below so doesn't seem necessary?

Not a big deal either way